File: src/main/java/com/server/ordering/domain/Order.java
package com.server.ordering.domain;

import com.server.ordering.domain.member.Customer;
import lombok.*;

import javax.persistence.*;
import java.time.ZonedDateTime;
import java.util.*;

import static com.server.ordering.domain.OrderStatus.*;
import static javax.persistence.CascadeType.*;
import static javax.persistence.EnumType.*;
import static javax.persistence.FetchType.*;
import static javax.persistence.GenerationType.*;
import static lombok.AccessLevel.*;

@Entity
@Getter @Setter
@Table(name = "orders")
@RequiredArgsConstructor
@NoArgsConstructor(access = PROTECTED)
public class Order {

    @Id @GeneratedValue(strategy = IDENTITY)
    @Column(name = "order_id")
    private Long id;

    @NonNull
    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "customer_id")
    private Customer customer;

    @OneToMany(mappedBy = "order", cascade = ALL, orphanRemoval = true)
    private final Set<OrderFood> orderFoods = new HashSet<>();

    private ZonedDateTime receivedTime;
    private ZonedDateTime checkedTime;
    private ZonedDateTime canceledOrCompletedTime;

    private Integer myOrderNumber;

    @OneToOne(mappedBy = "order")
    private Review review;

    @NonNull
    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "restaurant_id")
    private Restaurant restaurant;

    @NonNull
    @Enumerated(value = STRING)
    private OrderStatus status;
    @NonNull
    @Enumerated(value = STRING)
    private OrderType orderType;
    @NonNull
    private Integer tableNumber;
    @NonNull
    @Column(nullable = false)
    private Integer totalPrice;
    @NonNull
    private String orderSummary;

    public void addOrderFood(OrderFood orderFood) {
        this.orderFoods.add(orderFood);
        orderFood.registerOrder(this);
    }

    public static Order createOrder(Customer customer, Restaurant restaurant, List<OrderFood> orderFoods,
                                    OrderType orderType, Integer tableNumber, int totalPrice, String orderSummary) {
        Order order = new Order(customer, restaurant, ORDERED, orderType, tableNumber, totalPrice, orderSummary);
        orderFoods.forEach(order::addOrderFood);
        return order;
    }

    public Boolean isAbleToCancel() {
        return this.status == ORDERED;
    }

    public Boolean isAbleToOwnerCancel() {
        return this.status == ORDERED || this.status == CHECKED;
    }

    public Boolean isAbleToCheck() {
        return this.status == ORDERED;
    }

    public Boolean isAbleToComplete() {
        return this.status == CHECKED;
    }

    public void registerReview(Review review) {
        this.review = review;
    }

    public Boolean isAbleRegisterReview() {
        return Objects.isNull(this.review);
    }

    public void registerReceivedTime() {
        this.receivedTime = ZonedDateTime.now();
    }

    public void registerCheckedTime() { this.checkedTime = ZonedDateTime.now(); }

    public void registerCanceledTime() {
        this.canceledOrCompletedTime = ZonedDateTime.now();
    }

    public void registerCompletedTime() {
        this.canceledOrCompletedTime = ZonedDateTime.now();
    }

    public void cancel() {
        this.status = CANCELED;
    }

    public void check() { this.status = CHECKED; }

    public void complete() {
        this.status = COMPLETED;
    }

    public void removeAllOrderFood() {
        orderFoods.clear();
    }

    public void registerMyOrderNumber(int orderNumber) {
        this.myOrderNumber = orderNumber;
    }
}

File: src/main/kotlin/com/byeoru/ordering_server/domain/Order.kt
package com.byeoru.ordering_server.domain

import com.byeoru.ordering_server.domain.member.Customer
import java.time.ZonedDateTime
import java.util.*
import javax.persistence.*

@Entity
@Table(name = "orders")
class Order(customer: Customer,
            restaurant: Restaurant,
            status: OrderStatus,
            orderType: OrderType,
            tableNumber: Int,
            totalPrice: Int,
            orderSummary: String) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "order_id")
     var id: Long? = null
         protected set

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "customer_id")
     var customer: Customer = customer
         protected set

    @OneToMany(mappedBy = "order", cascade = [CascadeType.ALL], orphanRemoval = true)
     var orderFoods: MutableSet<OrderFood> = HashSet<OrderFood>()
         protected set

     var receivedTime: ZonedDateTime? = null
         protected set

     var checkedTime: ZonedDateTime? = null
         protected set

     var canceledOrCompletedTime: ZonedDateTime? = null
         protected set

    var myOrderNumber: Int? = null

    @OneToOne(mappedBy = "order")
     var review: Review? = null
         protected set

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "restaurant_id")
     var restaurant: Restaurant = restaurant
         protected set

    @Enumerated(value = EnumType.STRING)
     var status: OrderStatus = status
         protected set

    @Enumerated(value = EnumType.STRING)
     var orderType: OrderType = orderType
         protected set

     var tableNumber: Int = tableNumber
         protected set

    @Column(nullable = false)
     var totalPrice: Int = totalPrice
         protected set

     var orderSummary: String = orderSummary
         protected set

    fun addOrderFood(orderFood: OrderFood) {
        orderFoods.add(orderFood)
        orderFood.registerOrder(this)
    }

    val isAbleToCancel = status == OrderStatus.ORDERED
    val isAbleToOwnerCancel = status == OrderStatus.ORDERED || status == OrderStatus.CHECKED
    val isAbleToCheck = status == OrderStatus.ORDERED
    val isAbleToComplete = status == OrderStatus.CHECKED

    fun registerReview(review: Review?) {
        this.review = review
    }

    val isAbleRegisterReview: Boolean = Objects.isNull(review)

    fun registerReceivedTime() {
        receivedTime = ZonedDateTime.now()
    }

    fun registerCheckedTime() {
        checkedTime = ZonedDateTime.now()
    }

    fun registerCanceledTime() {
        canceledOrCompletedTime = ZonedDateTime.now()
    }

    fun registerCompletedTime() {
        canceledOrCompletedTime = ZonedDateTime.now()
    }

    fun cancel() {
        status = OrderStatus.CANCELED
    }

    fun check() {
        status = OrderStatus.CHECKED
    }

    fun complete() {
        status = OrderStatus.COMPLETED
    }

    fun removeAllOrderFood() {
        orderFoods.clear()
    }

    fun registerMyOrderNumber(orderNumber: Int) {
        myOrderNumber = orderNumber
    }

    companion object {
        @JvmStatic
        fun createOrder(
            customer: Customer, restaurant: Restaurant, orderFoods: List<OrderFood>,
            orderType: OrderType, tableNumber: Int, totalPrice: Int, orderSummary: String
        ): Order {
            val order =
                Order(customer, restaurant, OrderStatus.ORDERED, orderType, tableNumber, totalPrice, orderSummary)
            orderFoods.forEach(order::addOrderFood)
            return order
        }
    }
}